src/dag.rs

```diff
@@ -0,0 +1,483 @@
+    pub action: ActionName,
+
+    /// Each node may be the start of the outer saga, or a subsaga. We store
+    /// the creation parameters for the saga or subsaga as output of the first
+    /// node so that it can be retrieved by child nodes. In order to feed the
+    /// output of the node execution, the params themselves must be stored
+    /// as input to the DAG, and so we store them here. We only store these
+    /// parameters for the first node of a saga or child saga.
+    pub create_params: Option<serde_json::Value>,
+    // TODO:
+    // What if we hang a set of KV pairs or other structured data off each node
+    // which instructs that action what lookup keys to use? This could
+    // be instead of `instance_id`. The benefits are more information
+    // could be dynamically added to nodes in the DagBuilder. In essence, each node
+    // could get its own set of "create parameters".
+    // The downside is more work for creators of dags and users of actions. This
+    // would also likely be more error prone in that now each user of a dag
+    // has to add the right data structure to each node which runs an action or
+    // the action won't do the right thing. So instead of depending just on the output
+    // of node actions, the actions also depend on input of their current node.
+    // If we did this, could we typecheck these inputs with the actions? Maybe via
+    // some macro that generates a checker function.
+}
+
+/// A specification for creating a node. This allows dynamic creation of nodes
+/// for subsagas with different instance_ids and creation parameters.
+pub struct NodeSpec<'a> {
+    pub name: &'a str,
+    pub label: &'a str,
+    pub action: &'a str,
+}
+
+/// An abstract description of the shape of node execution
+///
+/// TODO: Do we want to allow arbitrary recursion here?
+pub enum NodeConcurrency<'a> {
+    Linear(Vec<NodeSpec<'a>>),
+    Parallel(Vec<NodeSpec<'a>>),
+}
+
+/// An abstract specification for a subsaga that can be used by a Dag
+/// to generate and append saga nodes.
+///
+/// An example implementation is shown below:
+///
+/// ```ignore
+/// fn create_disk_subsaga<'a>() -> SagaSpec<'a> {
+///     SagaSpec(
+///         // The root node is always the first one. It contains the saga
+///         // parameters.
+///         NodeSpec {
+///             name: "disk_create_params",
+///             label: "ReturnDiskCreateParams",
+///             action: "disk_create_params",
+///         },
+///         vec![
+///             NodeConcurrency::Linear(vec![NodeSpec {
+///                 name: "disk_id",
+///                 label: "GenerateDiskId",
+///                 action: "saga_generate_uuid",
+///             }]),
+///             NodeConcurrency::Parallel(vec![NodeSpec {...}, NodeSpec {...}]),
+///         ],
+///     )
+/// }
+/// ```
+///
+///
+/// TODO: Semantic validation (actions exist in registry, etc...)
+pub struct SagaSpec<'a>(pub NodeSpec<'a>, pub Vec<NodeConcurrency<'a>>);
```

Comment:
What if we do something like this:

```rust
pub enum DagSpec<'a> {
    Linear(Vec<DagSpec<'a>>),
    Parallel(Vec<DagSpec<'a>>),
    Node(NodeSpec<'a>),
}
```

We think of this is purely sugar for the DagBuilder ... maybe we move it to its own file. And the only operation we allow on it is `to_dag` as @davepacheco suggests:
https://github.com/oxidecomputer/steno/pull/30/files#diff-bc42300689f929cdb4146f0adf9bb529f0df6750b6a92bf396f50d256cf17489R480

I think this would help clarify that this is just a mechanism to structure Dag construction. We might even pull this out, force everyone to build dags by hand, and see if there are common patterns that this is going to simplify

Reply:
Great idea! That works for me.

Reply:
Based on recent changes that @davepacheco made, where DAGs can be added as subsagas with differently named param nodes, I don't believe specs make anything easier anymore. In fact, I can't figure out a way to make them expressive and allow different param node names for subsaga specs without passing in a spec and a tree of of param node names as separate arguments to `to_dag` that match the `DagSpec`. We'd pass them as separate parameters because different sagas may include a `DagSpec` twice and we'd want to allow the saga to choose the names of the nodes.

In light of this difficulty and the improvement in ergonomics of the `DagBuilder`, I'm going to drop the declarative spec idea altogether for now. We can always add something similar back in the future.